docs/adr/adr-013-bootstrap-from-previous-peers.md

```diff
@@ -0,0 +1,370 @@
+        }
+
+        for id, peer := range p.trackedPeers {
+                if peer.peerScore <= defaultScore {
+                        delete(p.trackedPeers, id)
+                }
+        }
++
++       updatedPeerList := make([]peer.ID, 0, len(p.trackedPeers))
++       for _, peer := range p.trackedPeers {
++           updatedPeerList = append(updatedPeerList, PIDtoAddrInfo(peer.peerID))
++       }
++
+        p.peerLk.Unlock()
++
++       p.onUpdatedPeers(updatedPeerList)
+                }
+        }
+ }
+```
+
+where are `OnBlockedPeer` will be called whenever `peerTracker` blocks a peer through its method `blockPeer`.
+
+```diff
++++ go-header/p2p/peer_tracker.go
+ // blockPeer blocks a peer on the networking level and removes it from the local cache.
+ func (p *peerTracker) blockPeer(pID peer.ID, reason error) {
+        // add peer to the blacklist, so we can't connect to it in the future.
+        err := p.connGater.BlockPeer(pID)
+        if err != nil {
+                log.Errorw("header/p2p: blocking peer failed", "pID", pID, "err", err)
+        }
+        // close connections to peer.
+        err = p.host.Network().ClosePeer(pID)
+        if err != nil {
+                log.Errorw("header/p2p: closing connection with peer failed", "pID", pID, "err", err)
+        }
+ 
+        log.Warnw("header/p2p: blocked peer", "pID", pID, "reason", reason)
+ 
+        p.peerLk.Lock()
+        defer p.peerLk.Unlock()
+        // remove peer from cache.
+        delete(p.trackedPeers, pID)
+        delete(p.disconnectedPeers, pID)
++       p.onBlockedPeer(PIDToAddrInfo(pID))
+ }
+```
+We are assuming a function named `PIDToAddrInfo` that converts a `peer.ID` to a `peer.AddrInfo` struct for this example's purpose. 
+
+The `peerTracker`'s constructor should be updated to accept these new event handlers:
+```diff
++++ go-header/p2p/peer_tracker.go
+ type peerTracker struct {
+ func newPeerTracker(
+        h host.Host,
+        connGater *conngater.BasicConnectionGater,
++       onUpdatedPeers func([]peer.ID),
++       onBlockedPeer func(peer.ID),
+ ) *peerTracker {
+```
+
+as well as the `libhead.Exchange`'s options and construction:
+
+```diff
++++ go-header/p2p/options.go
+ type ClientParameters struct {
+        // MaxHeadersPerRangeRequest defines the max amount of headers that can be requested per 1 request.
+        MaxHeadersPerRangeRequest uint64
+        // RangeRequestTimeout defines a timeout after which the session will try to re-request headers
+        // from another peer.
+        RangeRequestTimeout time.Duration
+        // TrustedPeersRequestTimeout a timeout for any request to a trusted peer.
+        TrustedPeersRequestTimeout time.Duration
+        // networkID is a network that will be used to create a protocol.ID
+        networkID string
+        // chainID is an identifier of the chain.
+        chainID string
++
++       onUpdatedPeers func([]peer.AddrInfo)
++       onBlockedPeer func(peer.AddrInfo)
```

Comment:
if we want to put it in params, it'd have to be accessible externally so needs to be exported.